### PR TITLE
feat: Add per-repo token support for private additional repos (fixes #151)

### LINKED ISSUE #151: feat: Add per-repo token support for private additional repos
-

### PR DESCRIPTION
**Pull request by myakove**
## Summary

Add optional per-repo token support for cloning private additional repositories, closing #151.

## Changes

- **Model**: Extended `AdditionalRepo` with optional `token` field (backward compatible)
- **Clone logic**: Added `_inject_token_into_url()` helper for HTTPS token-authenticated cloning; updated `clone_into()` to accept and use per-repo tokens
- **Security**: Tokens encrypted at rest, stripped from API responses, never logged
- **Encryption**: Added `_encrypt/_decrypt/_strip_additional_repos_tokens()` to `encryption.py`
- **CLI**: Supports `name:url@token` and `name:url:ref@token` syntax via `parse_additional_repos()`
- **Tests**: 30+ new tests across models, config, encryption, repository, and analyzer

## Deliverables

- [x] Code changes
- [x] Tests
- [x] Backward compatibility preserved

Fixes #151

## Summary by CodeRabbit

* **New Features**
  * Additional repositories may include optional authentication tokens (syntax: name:url@token); tokens are accepted, stored (encrypted) and used automatically when cloning private repos.

* **Bug Fixes**
  * Improved warnings for additional-repo clone failures to name the repo and error type.
  * Credentials embedded for cloning are scrubbed after use; resumption rejects unreadable/encrypted stored tokens.

* **Tests**
  * Expanded tests covering token parsing, encryption/decryption, cloning behavior, and edge cases.

### CODE REVIEW

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

File: src/jenkins_job_insight/encryption.py [hunk 1]
```diff
@@ -145,6 +145,79 @@ def _get_fernet() -> Fernet:
...
+def _strip_additional_repos_tokens(repos: list) -> list:
+    """Remove token fields from additional_repos entries for API responses."""
+    result = []
+    for entry in repos:
+        if isinstance(entry, dict):
+            stripped = {k: v for k, v in entry.items() if k != "token"}
+            result.append(stripped)
+        else:
+            result.append(entry)
+    return result
```
[the remark this answers is not in the document]
Comment (myakove) [reply src/jenkins_job_insight/encryption.py]: Extracted _map_additional_repo_entries helper to deduplicate the list traversal across all three functions. Each now passes only its entry-specific transform.

File: src/jenkins_job_insight/repository.py [hunk 2]
```diff
@@ -107,6 +107,24 @@ def _validate_repo_url(repo_url: str | HttpUrl) -> None:
...
+    from urllib.parse import urlparse, urlunparse
+
+    parsed = urlparse(url)
+    if parsed.scheme != "https" or not token:
+        return url
+    # Replace any existing userinfo with token-based auth
+    netloc = f"x-token-auth:{token}@{parsed.hostname}"
+    if parsed.port:
+        netloc += f":{parsed.port}"
+    return urlunparse(parsed._replace(netloc=netloc))
```
[the remark this answers is not in the document]
Comment (myakove) [reply src/jenkins_job_insight/repository.py:124]: Added quote(token, safe="") to percent-encode tokens before placing in URL userinfo. Also simplified host_port extraction using rsplit.

File: src/jenkins_job_insight/repository.py [hunk 3]
```diff
@@ -211,8 +231,11 @@ def clone_into(
         """
         _validate_repo_url(repo_url)
         target_dir.mkdir(parents=True, exist_ok=True)
+        clone_url = str(repo_url)
+        if token:
+            clone_url = _inject_token_into_url(clone_url, token)
         logger.info(f"Cloning repository into {target_dir}")
-        _clone_with_ssl_retry(str(repo_url), target_dir, depth, branch=branch)
+        _clone_with_ssl_retry(clone_url, target_dir, depth, branch=branch)
```
[the remark this answers is not in the document]
Comment (myakove) [reply src/jenkins_job_insight/repository.py]: Added post-clone credential scrub using git remote set-url to replace the token-bearing URL with the original clean URL in .git/config.

File: tests/test_config.py [hunk 4]
```diff
@@ -398,6 +398,49 @@ def test_duplicate_names_rejected(self) -> None:
...
+        assert "token" not in result[0]
+
+    def test_multiple_repos_mixed_tokens(self) -> None:
+        """Some repos with token, some without."""
+        result = parse_additional_repos(
+            "infra:https://github.com/org/infra@tok1,"
+            "product:https://github.com/org/product"
+        )
+        assert len(result) == 2
+        assert result[0]["token"] == "tok1"  # pragma: allowlist secret
```
[the remark this answers is not in the document]
Comment (myakove) [reply tests/test_config.py]: Added # noqa: S105 to suppress Ruff S105 for intentional test token literal.

File: tests/test_encryption.py [hunk 5]
```diff
@@ -261,3 +261,114 @@ def test_each_redacted_key_is_removed(self, key: str) -> None:
...
+        }
+        encrypted = encrypt_sensitive_fields(params)
+        decrypted = decrypt_sensitive_fields(encrypted)
+        assert (
+            decrypted["additional_repos"][0]["token"] == "tok1"
+        )  # pragma: allowlist secret
+        assert "token" not in decrypted["additional_repos"][1]
+        assert (
+            decrypted["additional_repos"][2]["token"] == "tok3"
+        )  # pragma: allowlist secret
```
[the remark this answers is not in the document]
Comment (myakove) [reply tests/test_encryption.py]: Added # noqa: S105 to all three token assertion lines (tok, tok1, tok3).

File: tests/test_repository.py [hunk 6]
```diff
@@ -669,3 +669,107 @@ def test_clone_no_branch_when_empty(self, mock_repo: MagicMock) -> None:
...
+    ) -> None:
+        """clone_into with token injects it into the clone URL."""
+        manager = RepositoryManager()
+        target = tmp_path / "my-repo"
+        manager.clone_into(
+            "https://github.com/org/repo",
+            target,
+            depth=1,
+            token="tok123",  # pragma: allowlist secret
+        )
```
[the remark this answers is not in the document]
Comment (myakove) [reply tests/test_repository.py:750]: Added # noqa: S106 to suppress Ruff S106 for intentional token keyword argument.

**Comment by myakove** [comment]
@coderabbitai[bot]

The following review comments were reviewed and a decision was made:

### `src/jenkins_job_insight/analyzer.py:88` (outside diff) — Avoid logging token-bearing clone exceptions.
> Addressed: Changed to log only repo name and exception type, preventing token-bearing URLs from appearing in log output.
### `src/jenkins_job_insight/config.py:57` (outside diff) — Do not echo token-bearing repo specs in parser errors.
> Addressed: Removed raw entry values from all four ValueError messages to prevent token-bearing repo specs from leaking in error output.

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

File: src/jenkins_job_insight/analyzer.py [hunk 7]
```diff
@@ -91,11 +91,16 @@ async def _clone_into_subdir(ar: AdditionalRepo) -> None:
...
             )
             cloned[ar.name] = target
             logger.info(f"Cloned additional repo '{ar.name}' into {target}")
         except Exception as e:
-            logger.warning(f"Failed to clone additional repo '{ar.name}': {e}")
+            logger.warning(
+                "Failed to clone additional repo '%s' (%s)",
+                ar.name,
+                type(e).__name__,
+            )
```
[the remark this answers is not in the document]
Comment (myakove) [reply src/jenkins_job_insight/analyzer.py]: Added # noqa: BLE001 with rationale comment to the intentional broad except for non-fatal clone failures.

File: src/jenkins_job_insight/encryption.py [hunk 8]
```diff
@@ -145,6 +146,71 @@ def _get_fernet() -> Fernet:
...
+                fernet = _get_fernet()
+            ciphertext = token_val[len(_ENCRYPTED_PREFIX) :]
+            try:
+                decrypted = fernet.decrypt(ciphertext.encode()).decode()
+                return {**entry, "token": decrypted}
+            except InvalidToken:
+                logger.warning(
+                    "Failed to decrypt additional_repos token: encryption key may have changed."
+                )
+        return entry
```
[the remark this answers is not in the document]
Comment (myakove) [reply src/jenkins_job_insight/encryption.py:201]: Added validation in _reconstruct_request_from_stored to check additional_repos tokens for undecryptable enc: prefix, failing fast on resume instead of passing corrupt tokens.

**Comment by myakove** [comment]
@coderabbitai[bot]

The following review comments were reviewed and a decision was made:

### `tests/test_config.py:411` (duplicate) — Add Ruff S105 suppression for these intentional token literals.
> Addressed: Added # noqa: S105 to both intentional token literal lines.
### `src/jenkins_job_insight/repository.py:237` (duplicate) — Scrub tokenized remotes on failure paths too.
> Addressed: Wrapped _clone_with_ssl_retry in try/finally to scrub credentials even on failure paths. Added .git existence check and shutil.which for safe git executable resolution.

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai[bot]

The following review comments were reviewed and a decision was made:

### `src/jenkins_job_insight/main.py:281` (outside diff) — Extract the repeated encrypted-value check.
> Addressed: Extracted _is_encrypted_value helper to deduplicate the enc: prefix check across both SENSITIVE_KEYS validation and additional_repos token validation.

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai review

**Comment by myakove** [comment]
@coderabbitai[bot]

The following review comments were reviewed and a decision was made:

### `src/jenkins_job_insight/analyzer.py:88` (outside diff) — Avoid logging token-bearing clone exceptions.
> Addressed: Already addressed: Changed to log only repo name and exception type.
### `src/jenkins_job_insight/config.py:57` (outside diff) — Do not echo token-bearing repo specs in parser errors.
> Addressed: Already addressed: Removed raw entry values from error messages.
### `tests/test_config.py:411` (duplicate) — Add Ruff S105 suppression for these intentional token literals.
> Addressed: Already addressed: Added # noqa: S105 to token literal lines.
### `src/jenkins_job_insight/repository.py:237` (duplicate) — Scrub tokenized remotes on failure paths too.
> Addressed: Already addressed: Wrapped in try/finally with .git check for credential scrubbing.
### `src/jenkins_job_insight/main.py:281` (outside diff) — Extract the repeated encrypted-value check.
> Addressed: Already addressed: Extracted _is_encrypted_value helper.

**Comment by myakove** [comment]
@coderabbitai review
